File: android/core/src/main/kotlin/com/poyal/drawry/core/Backup.kt
package com.poyal.drawry.core

import java.io.*
import java.nio.ByteBuffer
import java.security.MessageDigest
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString

@Serializable
data class BackupHeader(
    val version: Int = 2,
    val iterations: Int = 210_000,
    val salt: String,
    val passwordWrap: String,
    val recoveryWrap: String,
)

@Serializable data class BackupFile(val name: String, val length: Long, val sha256: String)

@Serializable data class BackupManifest(val snapshot: Snapshot, val files: List<BackupFile>)

object Backup {
    const val ITERATIONS = 210_000

    private fun b64(bytes: ByteArray) = Base64.getEncoder().encodeToString(bytes)

    private fun decode(text: String) = Base64.getDecoder().decode(text)

    fun passwordKey(password: String, salt: ByteArray): ByteArray {
        // Explicit UTF-8 avoids provider-dependent PBE character conversion.
        require(password.isNotEmpty())
        val mac =
            Mac.getInstance("HmacSHA256").apply {
                init(SecretKeySpec(password.toByteArray(Charsets.UTF_8), "HmacSHA256"))
            }
        var u = mac.doFinal(salt + ChunkCipher.uint32(1))
        val result = u.copyOf()
        repeat(ITERATIONS - 1) {
            u = mac.doFinal(u)
            for (i in result.indices) result[i] = (result[i].toInt() xor u[i].toInt()).toByte()
        }
        return result
    }

    private fun wrap(key: ByteArray, wrapping: ByteArray, label: String): String {
        val nonce = ChunkCipher.random(12)
        return b64(nonce + ChunkCipher.seal(key, wrapping, nonce, label.toByteArray()))
    }

    private fun unwrap(text: String, key: ByteArray, label: String): ByteArray {
        val box = decode(text)
        require(box.size == 60)
        return ChunkCipher.open(
            box.copyOfRange(12, box.size),
            key,
            box.copyOfRange(0, 12),
            label.toByteArray(),
        )
    }

    fun digest(bytes: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }

    fun digest(file: File): String {
        val hash = MessageDigest.getInstance("SHA-256")
        file.inputStream().buffered().use { input ->
            val bytes = ByteArray(1 shl 20)
            while (true) {
                val n = input.read(bytes)
                if (n < 0) break
                hash.update(bytes, 0, n)
            }
        }
        return hash.digest().joinToString("") { "%02x".format(it) }
    }

    fun create(
        snapshot: Snapshot,
        mediaDirectory: File,
        destination: File,
        password: String,
        recoveryKey: ByteArray,
    ) {
        require(password.codePointCount(0, password.length) >= 8 && recoveryKey.size == 32)
        snapshot.validate()
        val salt = ChunkCipher.random(16)
        val dataKey = ChunkCipher.random(32)
        val header =
            wireJson
                .encodeToString(
                    BackupHeader(
                        salt = b64(salt),
                        passwordWrap =
                            wrap(dataKey, passwordKey(password, salt), "drawry-password-v2"),
                        recoveryWrap = wrap(dataKey, recoveryKey, "drawry-recovery-v2"),
                    )
                )
                .toByteArray(Charsets.UTF_8)
        val files =
            snapshot.diaries
                .flatMap { it.media }
                .flatMap { it.files }
                .sorted()
                .map { name ->
                    val file = File(mediaDirectory, name)
                    require(file.isFile)
                    BackupFile(name, file.length(), digest(file))
                }
        val manifest =
            wireJson.encodeToString(BackupManifest(snapshot, files)).toByteArray(Charsets.UTF_8)
        require(manifest.size <= 32 * 1024 * 1024)
        val payload = File(destination.path + ".payload")
        try {
            // Open one media stream at a time, so diary count cannot exhaust descriptors.
            val reader =
                object : InputStream() {
                    var current: InputStream =
                        ByteArrayInputStream(ChunkCipher.uint32(manifest.size) + manifest)
                    var index = 0

                    override fun read(): Int {
                        val b = ByteArray(1)
                        return if (read(b, 0, 1) < 0) -1 else b[0].toInt() and 255
                    }

                    override fun read(b: ByteArray, off: Int, len: Int): Int {
                        if (len == 0) return 0
                        while (true) {
                            val n = current.read(b, off, len)
                            if (n >= 0) return n
                            current.close()
                            if (index == files.size) return -1
                            current =
                                File(mediaDirectory, files[index++].name).inputStream().buffered()
                        }
                    }

                    override fun close() {
                        current.close()
                    }
                }
            reader.use {
                ChunkCipher.encrypt(
                    it,
                    4L + manifest.size + files.sumOf { it.length },
                    payload,
                    dataKey,
                    digest(header),
                )
            }
            destination.outputStream().buffered().use { out ->
                out.write("DRBK2".toByteArray())
                out.write(ChunkCipher.uint32(header.size))
                out.write(header)
                payload.inputStream().use { it.copyTo(out) }
            }
        } catch (e: Exception) {
            destination.delete()
            throw e
        } finally {
            payload.delete()
            dataKey.fill(0)
        }
    }

    fun restore(
        source: File,
        staging: File,
        password: String? = null,
        recoveryKey: ByteArray? = null,
    ): Snapshot {
        require(!staging.exists() && staging.mkdirs())
        var success = false
        try {
            val payload = File(staging, "payload.tmp")
            val (headerBytes, key) =
                source.inputStream().buffered().use { input ->
                    require(input.readExactly(5).contentEquals("DRBK2".toByteArray())) {
                        "지원하지 않는 백업입니다."
                    }
                    val n = ByteBuffer.wrap(input.readExactly(4)).int
                    require(n in 1..65536)
                    val bytes = input.readExactly(n)
                    val header =
                        wireJson.decodeFromString<BackupHeader>(bytes.toString(Charsets.UTF_8))
                    val salt = decode(header.salt)
                    require(
                        header.version == 2 && header.iterations == ITERATIONS && salt.size == 16
                    )
                    val dataKey =
                        if (!password.isNullOrEmpty())
                            unwrap(
                                header.passwordWrap,
                                passwordKey(password, salt),
                                "drawry-password-v2",
                            )
                        else {
                            require(recoveryKey?.size == 32)
                            unwrap(header.recoveryWrap, recoveryKey!!, "drawry-recovery-v2")
                        }
                    payload.outputStream().buffered().use { input.copyTo(it) }
                    bytes to dataKey
                }
            val receiver = PayloadReceiver(staging)
            val snapshot =
                try {
                    ChunkCipher.decrypt(payload, key, digest(headerBytes), receiver::accept)
                    receiver.finish()
                } finally {
                    receiver.close()
                    key.fill(0)
                }
            payload.delete()
            snapshot.diaries
                .flatMap { it.media }
                .forEach { media ->
                    media.files.forEach { name ->
                        ChunkCipher.decrypt(File(staging, name), decode(media.key), name) {}
                    }
                }
            success = true
            return snapshot
        } finally {
            if (!success) staging.deleteRecursively()
        }
    }
}

private class PayloadReceiver(val directory: File) : Closeable {
    var buffer = ByteArray(0)
    var manifestLength: Int? = null
    var manifest: BackupManifest? = null
    var index = 0
    var written = 0L
    var output: OutputStream? = null

    fun accept(data: ByteArray) {
        buffer += data
        if (manifestLength == null && buffer.size >= 4) {
            manifestLength =
                ByteBuffer.wrap(buffer, 0, 4).int.also { require(it in 1..32 * 1024 * 1024) }
            buffer = buffer.copyOfRange(4, buffer.size)
        }
        val n = manifestLength
        if (manifest == null && n != null && buffer.size >= n) {
            manifest =
                wireJson
                    .decodeFromString<BackupManifest>(
                        buffer.copyOfRange(0, n).toString(Charsets.UTF_8)
                    )
                    .also { m ->
                        m.snapshot.validate()
                        require(
                            m.snapshot.diaries.flatMap { it.media }.flatMap { it.files }.sorted() ==
                                m.files.map { it.name }
                        )
                        require(m.files.all { it.length in 40..512L * 1024 * 1024 })
                    }
            buffer = buffer.copyOfRange(n, buffer.size)
        }
        val m = manifest ?: return
        var offset = 0
        while (index < m.files.size && offset < buffer.size) {
            val file = m.files[index]
            if (output == null) output = File(directory, file.name).outputStream().buffered()
            val count = minOf((buffer.size - offset).toLong(), file.length - written).toInt()
            output!!.write(buffer, offset, count)
            offset += count
            written += count
            if (written == file.length) {
                output!!.close()
                output = null
                require(Backup.digest(File(directory, file.name)) == file.sha256)
                index++
                written = 0
            }
        }
        buffer = buffer.copyOfRange(offset, buffer.size)
    }

    fun finish(): Snapshot {
        val m = requireNotNull(manifest)
        require(index == m.files.size && written == 0L && buffer.isEmpty())
        return m.snapshot
    }

    override fun close() {
        output?.close()
    }
}

File: ios/Core/Backup.swift
import CommonCrypto
import CryptoKit
import Foundation

public struct BackupHeader: Codable {
  public var version: Int
  public var iterations: Int
  public var salt: String
  public var passwordWrap: String
  public var recoveryWrap: String
}
public struct BackupFile: Codable {
  public var name: String
  public var length: UInt64
  public var sha256: String
}
public struct BackupManifest: Codable {
  public var snapshot: Snapshot
  public var files: [BackupFile]
}

public enum Backup {
  public static let iterations = 210_000
  public static func passwordKey(_ password: String, salt: Data) throws -> Data {
    let bytes = Array(password.utf8)
    var result = Data(count: 32)
    let status = result.withUnsafeMutableBytes { out in
      salt.withUnsafeBytes { s in
        bytes.withUnsafeBytes { p in
          CCKeyDerivationPBKDF(
            CCPBKDFAlgorithm(kCCPBKDF2), p.baseAddress!.assumingMemoryBound(to: Int8.self),
            bytes.count,
            s.baseAddress!.assumingMemoryBound(to: UInt8.self), salt.count,
            CCPseudoRandomAlgorithm(kCCPRFHmacAlgSHA256),
            UInt32(iterations), out.baseAddress!.assumingMemoryBound(to: UInt8.self), 32)
        }
      }
    }
    guard status == kCCSuccess else { throw VaultError.authentication }
    return result
  }
  private static func wrap(_ key: Data, with wrappingKey: Data, label: String) throws -> String {
    try AES.GCM.seal(key, using: SymmetricKey(data: wrappingKey), authenticating: Data(label.utf8))
      .combined!.base64EncodedString()
  }
  private static func unwrap(_ text: String, with key: Data, label: String) throws -> Data {
    guard let data = Data(base64Encoded: text), data.count == 60 else {
      throw VaultError.authentication
    }
    do {
      return try AES.GCM.open(
        AES.GCM.SealedBox(combined: data), using: SymmetricKey(data: key),
        authenticating: Data(label.utf8))
    } catch { throw VaultError.authentication }
  }
  public static func digest(_ url: URL) throws -> String {
    let file = try FileHandle(forReadingFrom: url)
    defer { try? file.close() }
    var hash = SHA256()
    while let data = try file.read(upToCount: 1 << 20), !data.isEmpty { hash.update(data: data) }
    return hash.finalize().map { String(format: "%02x", $0) }.joined()
  }
  public static func create(
    snapshot: Snapshot, mediaDirectory: URL, destination: URL, password: String, recoveryKey: Data
  ) throws {
    guard password.unicodeScalars.count >= 8, recoveryKey.count == 32 else {
      throw VaultError.invalidData
    }
    try snapshot.validate()
    let salt = try ChunkCipher.random(16)
    let dataKey = try ChunkCipher.random(32)
    let header = BackupHeader(
      version: 2, iterations: iterations, salt: salt.base64EncodedString(),
      passwordWrap: try wrap(
        dataKey, with: passwordKey(password, salt: salt), label: "drawry-password-v2"),
      recoveryWrap: try wrap(dataKey, with: recoveryKey, label: "drawry-recovery-v2"))
    let encodedHeader = try JSONEncoder().encode(header)
    let files = try snapshot.diaries.flatMap { $0.media.flatMap(\.files) }.sorted().map {
      name -> BackupFile in
      let url = mediaDirectory.appendingPathComponent(name)
      let length = try url.resourceValues(forKeys: [.fileSizeKey]).fileSize!
      return BackupFile(name: name, length: UInt64(length), sha256: try digest(url))
    }
    let manifest = try JSONEncoder().encode(BackupManifest(snapshot: snapshot, files: files))
    guard manifest.count <= 32 * 1024 * 1024 else { throw VaultError.invalidData }
    let prefix = ChunkCipher.uint32(UInt32(manifest.count)) + manifest
    let reader = SegmentReader(
      prefix: prefix, files: files.map { mediaDirectory.appendingPathComponent($0.name) })
    let temporary = destination.appendingPathExtension("payload")
    defer { try? FileManager.default.removeItem(at: temporary) }
    let context = SHA256.hash(data: encodedHeader).map { String(format: "%02x", $0) }.joined()
    try ChunkCipher.encrypt(
      length: UInt64(prefix.count) + files.reduce(0) { $0 + $1.length },
      destination: temporary, key: dataKey, context: context, read: reader.read)
    FileManager.default.createFile(atPath: destination.path, contents: nil)
    let out = try FileHandle(forWritingTo: destination)
    var success = false
    defer {
      try? out.close()
      if !success { try? FileManager.default.removeItem(at: destination) }
    }
    try out.write(
      contentsOf: Data("DRBK2".utf8) + ChunkCipher.uint32(UInt32(encodedHeader.count))
        + encodedHeader)
    let input = try FileHandle(forReadingFrom: temporary)
    defer { try? input.close() }
    while let bytes = try input.read(upToCount: 1 << 20), !bytes.isEmpty {
      try out.write(contentsOf: bytes)
    }
    try out.synchronize()
    success = true
  }
  /// Returns verified metadata; writes only encrypted media into a new staging directory.
  public static func restore(
    source: URL, staging: URL, password: String? = nil, recoveryKey: Data? = nil
  ) throws -> Snapshot {
    guard !FileManager.default.fileExists(atPath: staging.path) else {
      throw VaultError.invalidData
    }
    try FileManager.default.createDirectory(at: staging, withIntermediateDirectories: true)
    var success = false
    defer { if !success { try? FileManager.default.removeItem(at: staging) } }
    let input = try FileHandle(forReadingFrom: source)
    defer { try? input.close() }
    guard try input.readExactly(5) == Data("DRBK2".utf8) else { throw VaultError.unsupported }
    let length = Int(ChunkCipher.readUInt(try input.readExactly(4)))
    guard (1...65536).contains(length) else { throw VaultError.invalidData }
    let bytes = try input.readExactly(length)
    let header = try JSONDecoder().decode(BackupHeader.self, from: bytes)
    guard header.version == 2, header.iterations == iterations,
      let salt = Data(base64Encoded: header.salt), salt.count == 16
    else { throw VaultError.unsupported }
    let dataKey: Data
    if let password, !password.isEmpty {
      dataKey = try unwrap(
        header.passwordWrap, with: passwordKey(password, salt: salt), label: "drawry-password-v2")
    } else if let recoveryKey, recoveryKey.count == 32 {
      dataKey = try unwrap(header.recoveryWrap, with: recoveryKey, label: "drawry-recovery-v2")
    } else {
      throw VaultError.authentication
    }
    let payload = staging.appendingPathComponent("payload.tmp")
    FileManager.default.createFile(atPath: payload.path, contents: nil)
    let output = try FileHandle(forWritingTo: payload)
    do {
      while let part = try input.read(upToCount: 1 << 20), !part.isEmpty {
        try output.write(contentsOf: part)
      }
      try output.close()
    } catch {
      try? output.close()
      throw error
    }
    let receiver = PayloadReceiver(directory: staging)
    let context = SHA256.hash(data: bytes).map { String(format: "%02x", $0) }.joined()
    try ChunkCipher.decrypt(source: payload, key: dataKey, context: context, write: receiver.accept)
    let snapshot = try receiver.finish()
    try FileManager.default.removeItem(at: payload)
    // A valid archive must also contain valid media AEAD records, not merely matching checksums.
    for media in snapshot.diaries.flatMap(\.media) {
      for file in media.files {
        try ChunkCipher.decrypt(
          source: staging.appendingPathComponent(file), key: Data(base64Encoded: media.key)!,
          context: file
        ) { _ in }
      }
    }
    success = true
    return snapshot
  }
}

private final class SegmentReader {
  var prefix: Data
  var files: [URL]
  var handle: FileHandle?
  init(prefix: Data, files: [URL]) {
    self.prefix = prefix
    self.files = files
  }
  deinit { try? handle?.close() }
  func read(_ count: Int) throws -> Data {
    var result = Data()
    if !prefix.isEmpty {
      let n = min(count, prefix.count)
      result.append(prefix.prefix(n))
      prefix.removeFirst(n)
    }
    while result.count < count {
      if handle == nil {
        guard !files.isEmpty else { throw VaultError.invalidData }
        handle = try FileHandle(forReadingFrom: files.removeFirst())
      }
      let part = try handle!.read(upToCount: count - result.count) ?? Data()
      if part.isEmpty {
        try handle?.close()
        handle = nil
      } else {
        result.append(part)
      }
    }
    return result
  }
}

private final class PayloadReceiver {
  let directory: URL
  var buffer = Data()
  var manifestLength: Int?
  var manifest: BackupManifest?
  var index = 0
  var written: UInt64 = 0
  var output: FileHandle?
  init(directory: URL) { self.directory = directory }
  deinit { try? output?.close() }
  func accept(_ data: Data) throws {
    buffer.append(data)
    if manifestLength == nil, buffer.count >= 4 {
      let n = Int(ChunkCipher.readUInt(Data(buffer.prefix(4))))
      guard (1...32 * 1024 * 1024).contains(n) else { throw VaultError.invalidData }
      manifestLength = n
      buffer.removeFirst(4)
    }
    if manifest == nil, let n = manifestLength, buffer.count >= n {
      let m = try JSONDecoder().decode(BackupManifest.self, from: Data(buffer.prefix(n)))
      try m.snapshot.validate()
      let names = m.snapshot.diaries.flatMap { $0.media.flatMap(\.files) }.sorted()
      guard names == m.files.map(\.name),
        m.files.allSatisfy({ $0.length <= 512 * 1024 * 1024 && $0.length >= 40 })
      else { throw VaultError.invalidData }
      manifest = m
      buffer.removeFirst(n)
    }
    guard let manifest else { return }
    while index < manifest.files.count, !buffer.isEmpty {
      let file = manifest.files[index]
      if output == nil {
        let url = directory.appendingPathComponent(file.name)
        FileManager.default.createFile(atPath: url.path, contents: nil)
        output = try FileHandle(forWritingTo: url)
      }
      let n = Int(min(UInt64(buffer.count), file.length - written))
      try output!.write(contentsOf: buffer.prefix(n))
      buffer.removeFirst(n)
      written += UInt64(n)
      if written == file.length {
        try output?.close()
        output = nil
        guard try Backup.digest(directory.appendingPathComponent(file.name)) == file.sha256 else {
          throw VaultError.authentication
        }
        index += 1
        written = 0
      }
    }
  }
  func finish() throws -> Snapshot {
    guard let manifest, index == manifest.files.count, written == 0, buffer.isEmpty else {
      throw VaultError.invalidData
    }
    return manifest.snapshot
  }
}
